### PR TITLE
[FrameworkBundle] add a notice when passing a routerInterface without warmupInterface in RouterCacheWarmer

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yesish
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #23403
| License       | MIT
| Doc PR        | none

I'm adding a test to RouterCacheWarmer since there were none.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php [hunk 1]
```diff
@@ -43,6 +43,8 @@ public function warmUp($cacheDir)
     {
         if ($this->router instanceof WarmableInterface) {
             $this->router->warmUp($cacheDir);
+        } else {
+            trigger_error(sprintf('%s is not a instanceof %s, no warmUp has been triggered.', get_class($this->router), WarmableInterface::class), \E_USER_NOTICE);
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: IMHO we could trigger a deprecation now and throw a LogicException at the end.

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: Ddo we ever trigger a notice in the code base? This is suspicious to me :)

Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: We trigger warning but not notices you are right, I will change that with @chalasr's suggestion

Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: Since we are triggering a deprecation, we should go with master ?

Comment (chalasr) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: Yes, triggering/throwing does not fix a bug, it only enhances DX. This is for master in any case IMO

Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: That's for 4.1 IMHO.

Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: Rebase and throw added.

File: src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php [hunk 2]
```diff
@@ -38,14 +38,22 @@ public function __construct(ContainerInterface $container)
...
     public function warmUp($cacheDir)
     {
         $router = $this->container->get('router');
 
         if ($router instanceof WarmableInterface) {
             $router->warmUp($cacheDir);
+
+            return;
+        } else {
+            @trigger_error(sprintf('Passing a %s without implementing %s is deprecated since 4.1.', RouterInterface::class, WarmableInterface::class), \E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: I do not understand why we care about a deprecation if we immediately throw afterwards anyway.

Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: We MUST NOT throw (we cannot do it until 5.0). We should have only the deprecation notice.

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by Simperfit** [comment]
done @stof

**Comment by Simperfit** [comment]
I will update the Changelog

**Comment by Simperfit** [comment]
Travis failure seems unrelated

File: src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php [hunk 3]
```diff
@@ -45,6 +45,10 @@ public function warmUp($cacheDir)
 
         if ($router instanceof WarmableInterface) {
             $router->warmUp($cacheDir);
+
+            return;
+        } else {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: no need for the `else` as the previous `if` returns anyway

**Comment by Simperfit** [comment]
This has been updated @stof @xabbuh

**Comment by Simperfit** [comment]
Status: Needs Review

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 4]
```diff
@@ -5,7 +5,8 @@ CHANGELOG
 -----
 
  * allowed to pass an optional `LoggerInterface $logger` instance to the `Router`
-
+ * Passing a `RouterInterface` without `WarmableInterface` is deprecated.
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Using a `RouterInterface` that does not implement the `WarmableInterface` is deprecated and will not be supported in Symfony 5.0.

Comment (xabbuh) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Oh, and this must then be added to `UPGRADE-4.1.md` and `UPGRADE-5.0.md` (with a slight rewording of course) too.

**Comment by Simperfit** [comment]
@xabbuh done

File: UPGRADE-5.0.md [hunk 5]
```diff
@@ -5,7 +5,8 @@ Security
 --------
 
  * The `ContextListener::setLogoutOnUserChange()` method has been removed.
-
+ * Using a `RouterInterface` that does not implement the `WarmableInterface` is not supported anymore.
```
Comment (xabbuh) [inline UPGRADE-5.0.md]: This needs to be documented under `FrameworkBundle` not `Security`. :)

Comment (Simperfit) [reply UPGRADE-5.0.md]: done

File: UPGRADE-4.1.md [hunk 6]
```diff
@@ -16,3 +16,8 @@ Translation
 
  * The `FileDumper::setBackup()` method is deprecated and will be removed in 5.0.
  * The `TranslationWriter::disableBackup()` method is deprecated and will be removed in 5.0.
+
+FrameworkBundle
```
Comment (xabbuh) [inline UPGRADE-4.1.md]: Sorry for nitpicking @Simperfit, but could you please move this up sort entries alphabetically (makes searching for something much easier in the future with many more entries)?

Comment (Simperfit) [reply UPGRADE-4.1.md]: done

**Review by xabbuh — APPROVED** [review]

**Comment by chalasr** [comment]
Needs rebase

File: src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php [hunk 7]
```diff
@@ -45,7 +45,11 @@ public function warmUp($cacheDir)
 
         if ($router instanceof WarmableInterface) {
             $router->warmUp($cacheDir);
+
+            return;
         }
+
+        @trigger_error(sprintf('Passing a %s without implementing %s is deprecated since 4.1.', RouterInterface::class, WarmableInterface::class), \E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/CacheWarmer/RouterCacheWarmer.php]: since Symfony 4.1

**Comment by Simperfit** [comment]
PR rebased

**Comment by Simperfit** [comment]
PR Rebased

**Comment by Simperfit** [comment]
@nicolas-grekas this one is ready.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Simperfit.

File: UPGRADE-5.0.md [hunk 8]
```diff
@@ -11,11 +11,15 @@ EventDispatcher
 
  * The `TraceableEventDispatcherInterface` has been removed.
 
+FrameworkBundle
+---------------
+
+ * Using a `RouterInterface` that does not implement the `WarmableInterface` is not supported anymore.
```
Comment (Tobion) [inline UPGRADE-5.0.md]: This must mention that it is about RouterCacheWarmer. This has nothing to do with the RouterInterface itself.

**Pull request merged**